src/NHibernate.Test/TestCase.cs

```diff
@@ -154,30 +158,78 @@ protected virtual void OnTearDown()
+			var testResult = TestContext.CurrentContext.Result;
+			var fail = false;
+			string badCleanupMessage = null;
+			try
+			{
+				OnTearDown();
+				var wereClosed = CheckSessionsWereClosed();
+				var wasCleaned = CheckDatabaseWasCleaned();
+				var wereConnectionsClosed = CheckConnectionsWereClosed();
+				fail = !wereClosed || !wasCleaned || !wereConnectionsClosed;
 
-			bool wasClosed = CheckSessionWasClosed();
-			bool wasCleaned = CheckDatabaseWasCleaned();
-			bool wereConnectionsClosed = CheckConnectionsWereClosed();
-			bool fail = !wasClosed || !wasCleaned || !wereConnectionsClosed;
+				if (fail)
+				{
+					badCleanupMessage = "Test didn't clean up after itself. session closed: " + wereClosed + "; database cleaned: " + wasCleaned
+						+ "; connection closed: " + wereConnectionsClosed;
+					if (testResult != null && testResult.Outcome.Status == TestStatus.Failed)
+					{
+						// Avoid hiding a test failure (asserts are usually not hidden, but other exception would be).
+						badCleanupMessage = GetCombinedFailureMessage(testResult, badCleanupMessage, null);
+					}
+				}
+			}
+			catch (Exception ex)
+			{
+				if (testResult == null || testResult.Outcome.Status != TestStatus.Failed)
+					throw;
+
+				// Avoid hiding a test failure (asserts are usually not hidden, but other exceptions would be).
+				var exType = ex.GetType();
+				Assert.Fail(GetCombinedFailureMessage(testResult,
+					exType.Namespace + "." + exType.Name + " " + ex.Message,
+					ex.StackTrace));
+			}
 
 			if (fail)
 			{
-				Assert.Fail("Test didn't clean up after itself. session closed: " + wasClosed + " database cleaned: "+ wasCleaned
-					+ " connection closed: " + wereConnectionsClosed);
+				Assert.Fail(badCleanupMessage);
 			}
 		}
 
-		private bool CheckSessionWasClosed()
+		private string GetCombinedFailureMessage(ResultAdapter result, string tearDownFailure, string tearDownStackTrace)
+		{
+			var message = new StringBuilder()
+				.Append("The test failed and then failed to cleanup. Test failure is: ")
+				.AppendLine(result.Message)
+				.Append("Tear-down failure is: ")
+				.AppendLine(tearDownFailure)
+				.AppendLine("Test failure stack trace is: ")
+				.AppendLine(result.StackTrace);
+
+			if (!string.IsNullOrEmpty(tearDownStackTrace))
+				message.AppendLine("Tear-down failure stack trace is:")
+					.Append(tearDownStackTrace);
+
+			return message.ToString();
+		}
+
+		// Factory is exposed to descendant, any session opened directly from it will not be handled by following check.
+		private bool CheckSessionsWereClosed()
 		{
-			if (lastOpenedSession != null && lastOpenedSession.IsOpen)
+			var allClosed = true;
+			foreach (var session in _openedSessions)
 			{
-				log.Error("Test case didn't close a session, closing");
-				lastOpenedSession.Close();
-				return false;
+				if (session.IsOpen)
+				{
+					log.Error($"Test case didn't close session {(session as ISessionImplementor)?.SessionId}, closing");
```

Comment:
Huh, I was not expecting the internal contract to be available from public interface!